### ISSUE TITLE
Only One Subscriber Allowed Error After Migrating to Micronaut 4.2.x

### ISSUE THREAD

**Issue by RajeevMasseyTR** [micronaut-projects/micronaut-core#10328]
### Issue description

**Description**
After upgrading our project from Micronaut 3.8 to 4.2.x, we are encountering an issue where our HTTP client filter, specifically designed for retry logic with token authentication, is failing. The error we receive is "_**only one subscriber allowed**_". This issue was not present in the 3.8 version.

**Steps to Reproduce**

- Migrate a Micronaut project using HTTP client filters from version 3.8 to 4.2.x.
- Implement a custom HTTP client filter with retry logic for token authentication.
- Pass the invalid token in the Client Request
- Observe the behavior of the retry logic in the upgraded version.

**Expected Behavior**
The retry logic should work as it did in Micronaut 3.8, allowing for multiple retry attempts in case of an authentication failure.

**Actual Behavior**
The system throws an error stating java.lang.IllegalStateException: Only one subscriber allowed, preventing the retry logic from functioning as expected.

**Environment Information**
Micronaut Version: 4.2.x
JDK Version: 17

**Application Code**
A simplified version of the code causing the issue :

```
@RetryToken
@RequiredArgsConstructor
@Singleton
public class AuthFilter implements HttpClientFilter {
    private static final Logger LOGGER = LoggerFactory.getLogger(AuthFilter.class);
    private static final int ATTEMPTS = 2;

    // ... [other member variables]

    @Override
    public Publisher<? extends HttpResponse<?>> doFilter(MutableHttpRequest<?> request, ClientFilterChain chain) {
        LOGGER.info("Intercepted for auth setup");

        // ... [logic to determine target system and set up authentication]

        AtomicLong retryCount = new AtomicLong(0);

        return Flowable.fromPublisher(chain.proceed(request))
                .doOnSubscribe(subscription -> LOGGER.info("Processing request {}", request.getPath()))
                .retry(ATTEMPTS, (exThrowable) -> handleRetry(exThrowable, retryCount, request, targetSystem.get(), usedToken))
                .doOnNext(response -> LOGGER.info("Completed processing for request {}", request.getPath()))
                .doOnError(ex -> logFinalError(ex, retryCount))
                .subscribeOn(Schedulers.io());
    }

    // ... [additional methods such as handleRetry, logFinalError, etc.]
}
```

```

@FilterMatcher
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({ElementType.TYPE, ElementType.PARAMETER})
public @interface RetryToken {
}
```

```
@Client("${api.url}")
@RetryToken
public interface ExampleClient {
    @Get("/example/{param}")
    HttpResponse<ExampleResponse> getExample(@PathVariable String param);
}
```

**Additional Information**
This error seems to be related to changes in the way Micronaut handles HTTP requests or subscriptions in version 4.2.x compared to 3.8. Any guidance or workaround to resolve this issue would be greatly appreciated.

**Comment by graemerocher** [comment]
@yawkat should the request not be sent again if subscribed to again rather than subscribing to the same channel?

**Comment by RajeevMasseyTR** [comment]
@graemerocher Our goal is that the handleRetry method would refresh the token and updated the request token and then retry. I have tried @Retryable also and doesn't solve the purpose. So whatever do it will have to be in this filter

**Comment by yawkat** [comment]
the new filter infrastructure does not support executing downstream filters & controller multiple times like you want. it wouldn't really work anyway, eg because the request body can only be consumed once downstream anyway

**Comment by dstepanov** [comment]
@yawkat This is a client

**Comment by yawkat** [comment]
oh i see. it could work in theory there. but I don't think FilterRunner supports it and the downstream paths in the http client don't either.

**Comment by RajeevMasseyTR** [comment]
@yawkat So for now do we have a work round to achieve this

**Comment by yawkat** [comment]
no it is not possible to trigger multiple client requests in a client filter.

**Comment by RajeevMasseyTR** [comment]
@yawkat Is there any other way we can achieve the desired retry mechanism ?

**Comment by yawkat** [comment]
you need to do the retrying outside the DefaultHttpClient. But you excluded that above ("So whatever do it will have to be in this filter"), so I assume you have already considered this.

**Comment by graemerocher** [comment]
@RajeevMasseyTR have you considered using `@Retryable` instead of writing your own retry logic?

**Comment by RajeevMasseyTR** [comment]
@graemerocher I tried using it. But suppose if there is 401, I throw a particular error and in include the exception the retry didn't happen. 

```
@Retryable(attempts = "2", delay = "1s",
        includes = Exception.class)
```

Do i need something else as well

**Comment by yawkat** [comment]
where did you put the Retryable?

**Comment by RajeevMasseyTR** [comment]
@yawkat . In the HTTP client

```
@Client("${api.url}")
@Retryable(attempts = "2", delay = "1s",
        includes = Exception.class)@Client("${api.url}")
@RetryToken
public interface ExampleClient {
    @Get("/example/{param}")
    HttpResponse<ExampleResponse> getExample(@PathVariable String param);
}
```

**Comment by graemerocher** [comment]
you can use a `RetryPredicate` to solve this

**Comment by RajeevMasseyTR** [comment]
@graemerocher So in place of includes  i shall use predicate in @Retryable and then create another  error class extending RetryPredicate ? 

Also is it possible to do it in version 4.2.1

**Comment by graemerocher** [comment]
yes `@Retryable` is supported in 4.2.1 you must add the `micronaut-retry` dependency to use it

**Comment by RajeevMasseyTR** [comment]
@graemerocher I have implemented the @Retryable annotation. But the problems is its retrying the same request thrice with the same bearer token.

```
 return Flowable.fromPublisher(chain.proceed(request))
                .doOnError(ex -> {
                    if (FilterUtils.isUnauthorizedStatusCodePresent2(ex)) {
                        refreshTokenLogic(request);
                    }
                })
                .subscribeOn(Schedulers.io());
    }
```

now after one failed attempt we want to run the refreshTokenLogic with the http request. and when RetryPredicate is extended the http request is not present

**Comment by srinunagulapalli** [comment]
@yawkat I see the issue is marked with Improvement tag. 
Is this being considered to be implemented in the further releases? If yes, any approximate timelines?

**Comment by yawkat** [comment]
Potentially it will happen but there's no timeline.

**Comment by graemerocher** [comment]
Triage: Type **Bug** / `type: bug`; this looks like a client-filter retry/resubscription problem rather than a general enhancement. The simplified report does not include a current runnable reproducer, so please retest on 5.2.x with the smallest filter that retries an unauthorized response. The shown `Flowable.fromPublisher(chain.proceed(request)).retry(...)` subscribes to the same publisher again; because the current client path contains one-subscriber publishers, the filter should likely use `Flowable.defer(() -> chain.proceed(request))` and recreate or safely replay any one-shot request state on each attempt. If that still fails with `defer`, the client filter contract should be made explicit or made retry-safe, with a regression test.
